exp/addons/controllers/clusterresourceset_controller_test.go

```diff
@@ -292,21 +292,18 @@ metadata:
 			},
 			Data: map[string]string{},
 		}
+
+		// Let's wait until the initial reconciliations are *all* done, so we can test
+		// if the ConfigMap creation triggers a reconciliation. Otherwise a previous change
+		// of another resource triggers the reconciliation and we cannot verify if we react
+		// correctly to the ConfigMap creation
+		time.Sleep(5 * time.Second)
```

Comment:
This has been added to ensure the test would consistently fail without the fix in the ClusterRersourceSet controller.

Reply:
Short of wrapping the reconciler to count reconciles (which I would recommend avoiding), I don't think there's really anything that can be done. You also need to account for other controllers and tests in the test suite that may also be creating events, so to get the queue to truly be empty will be very very difficult AFAIK.

If we wanted to be more certain that, we could potentially create a new test with a separate manager watching a different test namespace (I'm not sure this is actually possible here, this is more general advice), that would then be more isolated. In these scenarios it's easier to verify reaction to specific events occurring.

Out of interest, without this, were you seeing the test flaking at all?

Reply:
not ideal but could think about something like adding an optional event recorder to the `resourceToClusterResourceSet` signature that we only pass through in tests that would record the exact resource and operation triggering the reconciliation

Reply:
What about relying on Observed generation instead?

Reply:
@JoelSpeed 
The problem is that without the timeout the test was not reliably able to detect the broken code. So this means when I drop the timeout the test has a certain probability that it doesn't test the reconcile of the create ConfigMap but one of the previous one either Cluster or ClusterResourceSet creation. So this only makes the test "greener" but just because we're testing the wrong thing. So the test was flaky because it couldn't reliably detect the broken code, with the timeout it can.

@enxebre Would it be possible with this solution to detect which exact reconcile of which event lead to the condition we're checking in l.313? If it was already done by a previous reconcile and the "right" reconcile does a no-op, the test is already testing the wrong thing.

@fabriziopandini Do you mean instead of the sleep I should wait until the resources has a specific ObservedGeneration and only after that create the ConfigMap? I'm not entirely sure which reconcilation is currently the problem. I suspect there might be one which is actually a no-op. I'm not sure if I can check for that one when it does nothing :/.

Reply:
@fabriziopandini I added the following here: https://github.com/kubernetes-sigs/cluster-api/blob/2ff65de19a19e018c5ba7fb75376a858861fc625/exp/addons/controllers/clusterresourceset_controller.go#L117
```golang
fmt.Printf("reconcile: %v %v\n", req.NamespacedName, clusterResourceSet.ObjectMeta.Generation)
```
output: (all before `testEnv.Create`):
```bash
reconcile: default/clusterresourceset-pug52m 1
reconcile: default/clusterresourceset-pug52m 1
reconcile: default/clusterresourceset-pug52m 1
reconcile: default/clusterresourceset-pug52m 2
```

So think you're right. I can just wait until `observedGeneration`: `2` instead of the sleep. I was suspecting the last reconcile triggers another one which doesn't change the resource, but apparently I was wrong.

WDYT?

Reply:
@sbueringer 
Catching up after lagging a little bit on this...

I think we should avoid to add a sleep here, because ensuring that watches with partial metadata works as expected is not really something a behavioural test for ClusterResourceSet should be concerned of.

If we want to have test coverage ensures that watches with partial metadata works well, I think we should have a dedicated test, might be under `util/predicates` so we can separate this issue from the CRS specific tests. 
This will allow us to use a dedicated, super simple, test reconciler that e.g. increases an annotation value at every reconcile, so we can actually check when a reconcile happens.

Does this makes sense? This can be eventually addressed also in a separated PR...